src/Utilities/copy.jl

```diff
@@ -42,449 +42,377 @@ end
-            filter!(filter_constraints, cis_src)
-        end
-        # do the rest in `pass_cons` which is type stable
-        pass_cons(dest, src, idxmap, cis_src)
-
-        cis_src =
-            MOI.get(src, MOI.ListOfConstraintIndices{MOI.SingleVariable,S}())
-        if filter_constraints !== nothing
-            filter!(filter_constraints, cis_src)
-        end
-        pass_attributes(dest, src, copy_names, idxmap, cis_src, pass_attr)
+    # Copy MOI.SingleVariable and MOI.VectorOfVariables constraints.
+    for cache in variable_constraint_cache
+        _pass_constraint_cache(
+            dest,
+            src,
+            copy_names,
+            index_map,
+            cache,
+            pass_constraints,
+            pass_attributes,
+        )
     end
-
-    for (S, cis_src) in
-        zip(vector_of_variables_types, vector_of_variables_indices)
-        if filter_constraints !== nothing
-            filter!(filter_constraints, cis_src)
+    # Copy remaining constraint types.
+    for (F, S) in MOI.get(src, MOI.ListOfConstraintTypesPresent())
+        if F == MOI.SingleVariable || F == MOI.VectorOfVariables
+            continue
         end
-        # do the rest in `pass_cons` which is type stable
-        pass_cons(dest, src, idxmap, cis_src)
-
-        cis_src =
-            MOI.get(src, MOI.ListOfConstraintIndices{MOI.VectorOfVariables,S}())
-        if filter_constraints !== nothing
-            filter!(filter_constraints, cis_src)
-        end
-        pass_attributes(dest, src, copy_names, idxmap, cis_src, pass_attr)
-    end
-
-    nonvariable_constraint_types = [
-        (F, S) for
-        (F, S) in MOI.get(src, MOI.ListOfConstraintTypesPresent()) if
-        F != MOI.SingleVariable && F != MOI.VectorOfVariables
-    ]
-    pass_nonvariable_constraints(
-        dest,
-        src,
-        idxmap,
-        nonvariable_constraint_types,
-        pass_cons;
-        filter_constraints = filter_constraints,
-    )
-    for (F, S) in nonvariable_constraint_types
         cis_src = MOI.get(src, MOI.ListOfConstraintIndices{F,S}())
         if filter_constraints !== nothing
             filter!(filter_constraints, cis_src)
         end
-        # do the rest in `pass_cons` which is type stable
-        pass_attributes(dest, src, copy_names, idxmap, cis_src, pass_attr)
-    end
-    return
-end
-
-function copy_free_variables(
-    dest::MOI.ModelLike,
-    idxmap::IndexMap,
-    vis_src,
-    copy_variables::F,
-) where {F<:Function}
-    if length(vis_src) != length(idxmap.var_map)
-        vars = copy_variables(dest, length(vis_src) - length(idxmap.var_map))
-        i = 1
-        for vi in vis_src
-            if !haskey(idxmap, vi)
-                idxmap[vi] = vars[i]
-                i += 1
-            end
-        end
-        @assert i == length(vars) + 1
-        @assert length(vis_src) == length(idxmap.var_map)
+        pass_constraints(dest, src, index_map, cis_src)
```

Comment:
@blegat in this PR I replaced `pass_nonvariable_constraints` with this call to `pass_constraints`. However given https://github.com/jump-dev/MathOptInterface.jl/issues/1341#issuecomment-840265697, the designs seem to be going in opposite directions, so I'll hold off on this PR until MatrixOfConstraints is finished.

At least we know the current design of `copy` isn't the inference bottleneck